### PR TITLE
hwdef: Add AP_HW_JHEMCUGF16F405 board config

### PR DESCRIPTION
**Pull request by alex-pokydin**
Add support for hardware
https://www.jhemcu.com/e_productshow/?76-JHEMCU-GF16-BMI-stack-F405-Flight-control-13A-ESC-76.html

### CODE REVIEW

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
needs readme file with information like this one:https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_ChibiOS/hwdef/SpeedyBeeF405WING/Readme.md

once created,I will review again

**Comment by alex-pokydin** [comment]
> needs readme file with information like this one:https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_ChibiOS/hwdef/SpeedyBeeF405WING/Readme.md
> 
> once created,I will review again

@Hwurzburg , done, please, review the PR,
thanks!

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:54]:
```suggestion
|SERIAL2|TX2/RX2|USART2 (DMA RX only) - User |
```

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:55]:
```suggestion
|SERIAL3|TX3/RX3|USART3 (DMA TX only) - User, GPS |
```

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:56]:
```suggestion
|SERIAL4|TX4/RX4|UART4 (DMA TX only) - User, VTX |
```

if you really intend this to be for VTX control, either as SmartAudio or DJI or Display Port, it will need a hwdef define to set the protocol...otherwise its GPS by default

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:53]: I would add a define to set the default to RCIN protocol in hwdef

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:57]:
```suggestion
|SERIAL6|TX6/RX6|USART6 -  User|
```

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:61]: SBUS inout must be on SBUS pin, all others can use RX1, CRSF/ELRS/SRXL2 must also use TX pin

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
also, should have only two commits
one for the bootloaders in Tools and one for hwdef in AP_HAL_ChibiOS

**Comment by alex-pokydin** [comment]
@Hwurzburg , thank you for your corrections.
I do my best to fix/cleanup all points
Please, check if all changes are correct
Thank you!

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:56]:
```suggestion
|SERIAL4|TX4/RX4|UART4 (DMA TX only) - GPS2 |
```

Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:56]: and you did not adddress my previous comment...based on your wiring diagram you expect users to use this port for VTX control...you should set the protocol in the hwdef to DEFAULT_SERIAL4_PROTOCOL SerialProtocol_Tramp or _SmartAudio

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:61]:
```suggestion
SBUS input must be on SBUS pin, all others can use RX1, CRSF/ELRS/SRXL2 must also use TX pin
```

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md]: sorry for my typo

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
a few things were not addressed

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (alex-pokydin) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:54]: @Hwurzburg , could you, please, explain the reason of this change?

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (alex-pokydin) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:55]: could you, please, explain the reason of this change?

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (alex-pokydin) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:56]: can we keep the port just as a general purpose port, so users will decide how they will use it?

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:55]: because the third and fourth uart protocols are set as GSP by default....and I checked the DMA assignments in the build configuration after waf configure and only those pins got DMA channels assigned

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:54]: see below

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md]: you can but BF users using your wiring diagram will have a harder time transitioning to AP

**Comment by alex-pokydin** [comment]
> a few things were not addressed

@Hwurzburg , sorry again)
I've just added VTX to hwdef and added your corrections to uarts
please, check if it is ok now
Thank you!

**Review by Hwurzburg — APPROVED** [review]
looks good...you might want to contact AndyPiper and see if you can get BIDIR setup on at least the first four outputs...your choice

**Comment by alex-pokydin** [comment]
> looks good...you might want to contact AndyPiper and see if you can get BIDIR setup on at least the first four outputs...your choice

Thank you!
BIDIR - it depends on ESC?

**Comment by Hwurzburg** [comment]
yes....but a user may change from the bundled one or you may decide to upgrade the one you supply...if you dont include the capabiliiy, you cant have anyone migrate in the future without a firmware change...it has no downside to to include it...IF it can be done without removing DMAs from UART1 which you need for CRSF...your choice

**Comment by alex-pokydin** [comment]
> looks good...you might want to contact AndyPiper and see if you can get BIDIR setup on at least the first four outputs...your choice

@andyp1per , could you, please, help me with BIDIR ? How can I check it and set it up?
Thank you!

**Comment by Hwurzburg** [comment]
@alex-pokydin I have again asked @andyp1per to look at this PR to suggest how to add bidir

**Comment by alex-pokydin** [comment]
@Hwurzburg , lets keep it as it is for now and I'll add bidir in the next iteration if it is possible
what do you think?

**Comment by Hwurzburg** [comment]
your call I''l add a devcall label

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef-bl.dat
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef-bl.dat:10]: can use AP_HW_JHEMCUGF16F405 instead of the numebr

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef.dat
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef.dat:12]: use AP_HW_JHEMCUGF16F405 not the number

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/JHEF-JHEF405PRO.config
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/JHEF-JHEF405PRO.config:3]: i don't think we want to include the betaflight config in our tree

**Review by tridge — CHANGES_REQUESTED** [review]
the commit messages need fixing, so "hwdef: added JHEMCUGF16F405"
and "Tools: added JHEMCUGF16F405 bootloader"

**Comment by tridge** [comment]
@alex-pokydin if you have trouble with the git commit fixups then I can do it for you
ping me in code_review channel

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef-bl.dat
Comment (alex-pokydin) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef-bl.dat]: fixed

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef.dat
Comment (alex-pokydin) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef.dat]: fixed

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/JHEF-JHEF405PRO.config
Comment (alex-pokydin) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/JHEF-JHEF405PRO.config]: fixed

**Comment by alex-pokydin** [comment]
> @alex-pokydin if you have trouble with the git commit fixups then I can do it for you ping me in code_review channel

Hello @tridge , 
fixed, please, check it
thanks!

**Review by tridge — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md]: still there telem is no the default protocol for usart6....I am approving anyway and will fix when I do the wiki page

**Pull request merged**
